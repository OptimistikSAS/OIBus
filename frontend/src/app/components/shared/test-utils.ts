import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };
