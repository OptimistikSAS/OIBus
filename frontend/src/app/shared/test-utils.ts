import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';
import { Page } from './types';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };

export function toPage<T>(content: Array<T>, totalElements: number = content.length, num = 0, size = Math.max(20, totalElements)): Page<T> {
  return {
    content,
    totalElements,
    number: num,
    size,
    totalPages: Math.ceil(totalElements / size)
  };
}

export function emptyPage<T>(size = 20): Page<T> {
  return {
    totalElements: 0,
    content: [] as Array<T>,
    size,
    totalPages: 0,
    number: 0
  };
}
