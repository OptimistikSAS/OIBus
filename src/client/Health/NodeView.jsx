import React from 'react'
import PropTypes from 'prop-types'
import { Row, Container, Button, UncontrolledTooltip } from 'reactstrap'
import { Link } from 'react-router-dom'
import ReactFlow from 'react-flow-renderer'
import { ConfigContext } from '../context/configContext.jsx'
import PointsButton from '../South/PointsButton.jsx'
import Modal from '../components/Modal.jsx'

const colors = {
  border: {
    enabled: '1px solid #2ea948',
    disabled: '1px solid #eaecef',
    warning: '1px solid #ffc107',
    success: '1px solid #2ea948',
  },
  background: {
    enabled: '#e1ffe15c',
    disabled: '#eaecef5c',
    warning: '#eaecef5c',
    success: '#e1ffe15c',
  },
}
const height = 500
const width = 1240

const NodeView = ({ status, onRestart, onShutdown }) => {
  const { activeConfig } = React.useContext(ConfigContext)
  const applications = activeConfig?.north?.applications
  const dataSources = activeConfig?.south?.dataSources
  const engine = activeConfig?.engine

  const northNodes = applications?.map((application, index) => (
    {
      id: application.applicationId, // unique id of node
      type: 'output', // output node
      targetPosition: 'bottom', // handle is at the bottom
      style: {
        background: colors.background[application.enabled ? 'enabled' : 'disabled'],
        border: colors.border[application.enabled ? 'enabled' : 'disabled'],
        width: '140px',
      },
      data: {
        label: (
          <div id={`north-${index}`} className={`oi-box tight text-${application.enabled ? 'success' : 'muted'}`}>
            <Link to={`/north/${application.applicationId}`}>
              <div style={{ overflow: 'hidden', textOverflow: 'ellipsis' }}>
                {application.applicationId}
              </div>
              <UncontrolledTooltip placement="top" target={`north-${index}`} innerClassName="oi-pop">
                {application.applicationId}
              </UncontrolledTooltip>
              <div>{`(${application.api})`}</div>
            </Link>
          </div>),
      },
      // position the node with an offset to center and then an offset for each node
      position: { x: 620 - 75 * applications.length + index * 150, y: 0 },
    }
  )) ?? []
  const northLinks = applications?.map((application) => (
    {
      id: `${application.applicationId}-engine`,
      source: 'engine',
      target: application.applicationId,
      animated: true,
      type: 'default',
      arrowHeadType: 'arrow',
      isHidden: !application.enabled,
    }
  )) ?? []

  const southNodes = dataSources?.map((dataSource, index) => (
    {
      id: dataSource.dataSourceId,
      type: 'input',
      sourcePosition: 'top',
      style: {
        background: colors.background[dataSource.enabled ? 'enabled' : 'disabled'],
        border: colors.border[dataSource.enabled ? 'enabled' : 'disabled'],
        width: '140px',
      },
      data: {
        label: (
          <div id={`south-${index}`} className={`oi-box tight text-${dataSource.enabled ? 'success' : 'muted'}`}>
            <Link to={`/south/${dataSource.dataSourceId}`}>
              <div style={{ overflow: 'hidden', textOverflow: 'ellipsis' }}>
                {dataSource.dataSourceId}
              </div>
              <UncontrolledTooltip placement="top" target={`south-${index}`} innerClassName="oi-pop">
                {dataSource.dataSourceId}
              </UncontrolledTooltip>
              <div>{`(${dataSource.protocol})`}</div>
            </Link>
            <PointsButton dataSource={dataSource} />
          </div>
        ),
      },
      // postion the node with an offset to center and then an offset for each node
      // 6 per line max => potentially render on several lines with y
      position: { x: 620 - (75 * Math.min(dataSources.length, 8)) + (index % 8) * 150, y: 250 + 120 * Math.trunc(index / 8) },
    }
  )) ?? []

  const southLinks = dataSources?.map((dataSource) => (
    {
      id: `${dataSource.dataSourceId}-engine`,
      target: 'engine',
      source: dataSource.dataSourceId,
      animated: true,
      type: 'default',
      arrowHeadType: 'arrow',
      isHidden: !dataSource.enabled,
    }
  )) ?? []

  const elements = [
    ...northNodes,
    ...northLinks,
    {
      id: 'engine',
      data: {
        label: (
          <div className="oi-box">
            <Link to="/engine">
              <div className={`text-${engine?.safeMode ? 'warning oi-safe-mode' : 'success'} center`}>
                {`Engine ${status?.version}${engine?.safeMode ? ' - Safe mode warning' : ''}`}
              </div>
            </Link>
            <Modal show={false} title="Server restart" body="Confirm restart?">
              {(confirm) => (
                <Button
                  className="inline-button autosize oi-restart-button"
                  color={engine?.safeMode ? 'warning' : 'success'}
                  onClick={confirm(onRestart)}
                  size="sm"
                  outline
                >
                  Restart
                </Button>
              )}
            </Modal>
            <Modal show={false} title="Server shutdown" body="Confirm shutdown?">
              {(confirm) => (
                <Button
                  className="inline-button autosize oi-shutdown-button"
                  color={engine?.safeMode ? 'warning' : 'success'}
                  onClick={confirm(onShutdown)}
                  size="sm"
                  outline
                >
                  Shutdown
                </Button>
              )}
            </Modal>
          </div>
        ),
      },
      position: { x: 20, y: 125 },
      targetPosition: 'bottom',
      sourcePosition: 'top',
      style: {
        background: colors.background[engine?.safeMode ? 'warning' : 'success'],
        color: 'black',
        border: colors.border[engine?.safeMode ? 'warning' : 'success'],
        width: 1080,
      },
    },
    {
      id: 'alive',
      type: 'input',
      sourcePosition: 'left',
      data: {
        label: (
          <Link to="/engine">
            <div
              className={`oi-box d-flex align-items-center text-${engine?.aliveSignal?.enabled ? 'success' : 'muted'}`}
            >
              <div className="oi-alive d-flex align-items-center">
                Alive
              </div>
            </div>
          </Link>
        ),
      },
      position: { x: 1100, y: 125 },
      style: {
        background: colors.background[engine?.aliveSignal?.enabled ? 'enabled' : 'disabled'],
        border: colors.border[engine?.aliveSignal?.enabled ? 'enabled' : 'disabled'],
        width: 100,
      },
    },
    ...southNodes,
    ...southLinks,
  ]

  const onLoad = (reactFlowInstance) => {
    reactFlowInstance.setTransform({ y: 0, x: 0, zoom: 0.8 })
  }

  return (
    <Container>
      <Row>
        <div style={{ height, width }}>
          <ReactFlow
            elements={elements}
            zoomOnScroll={false}
            nodesConnectable={false}
            elementsSelectable
            nodesDraggable={false}
            onLoad={onLoad}
          />
        </div>
      </Row>
    </Container>
  )
}

NodeView.propTypes = {
  status: PropTypes.object,
  onRestart: PropTypes.func,
  onShutdown: PropTypes.func,
}
NodeView.defaultProps = {
  status: {},
  onRestart: () => null,
  onShutdown: () => null,
}

export default NodeView
