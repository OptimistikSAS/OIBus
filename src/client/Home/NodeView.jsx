import React from 'react'
import PropTypes from 'prop-types'
import { Container } from 'reactstrap'
import ReactFlow from 'react-flow-renderer'
import { Link } from 'react-router-dom'
import { ConfigContext } from '../context/ConfigContext.jsx'
import { AlertContext } from '../context/AlertContext.jsx'
import ApiSchemas from '../North/Apis.jsx'
import ProtocolSchemas from '../South/Protocols.jsx'
import OIbCheckBox from '../components/OIbForm/OIbCheckBox.jsx'
import NorthMenu from './NorthMenu.jsx'
import SouthMenu from './SouthMenu.jsx'
import EngineMenu from './EngineMenu.jsx'
import imageCategories from './imageCategories'

const NODE_BACKGROUND_COLOR = '#eaecef5c'
const NODE_BORDER = '1px solid #eaecef'

const NodeView = ({
  onRestart,
  onShutdown,
}) => {
  const {
    newConfig,
    dispatchNewConfig,
  } = React.useContext(ConfigContext)
  const [connectorData, setConnectorData] = React.useState({})
  const [sseSource, setSseSource] = React.useState(null)
  const { setAlert } = React.useContext(AlertContext)

  const onLoad = (reactFlowInstance) => {
    reactFlowInstance.setTransform({
      y: 35,
      x: 0,
      zoom: 0.9,
    })
  }
  const onChange = (name, value, validity) => {
    dispatchNewConfig({
      type: 'update',
      name,
      value,
      validity,
    })
  }

  React.useEffect(() => {
    const source = new EventSource('/engine/sse')
    source.onerror = (error) => {
      setAlert({
        text: error.message,
        type: 'danger',
      })
    }
    source.onmessage = (event) => {
      if (event && event.data) {
        const myData = JSON.parse(event.data)
        setConnectorData(myData)
      }
    }
    setSseSource(source)

    return (() => {
      if (sseSource) {
        sseSource.close()
      }
    })
  }, [])

  const engineName = newConfig?.engine?.engineName ?? ''
  const applications = newConfig?.north?.applications ?? []
  const dataSources = newConfig?.south?.dataSources ?? []

  const northNodes = applications.map((application, indexNorth) => (
    {
      id: application.id,
      type: 'output',
      targetPosition: 'bottom',
      style: {
        background: NODE_BACKGROUND_COLOR,
        border: NODE_BORDER,
        width: 180,
        height: 130,
        padding: 0,
        borderRadius: 5,
      },
      data: {
        label: (
          <div className="d-flex flex-column h-100 w-100">
            <div className="w-100 d-flex flex-row justify-content-between align-items-center p-1 oi-node-header">
              <img
                src={`${imageCategories[ApiSchemas[application.api].category].image}` ?? imageCategories.Default.image}
                alt="logo"
                height="25px"
              />
              <div className="oi-node-title">
                {`${application.name}`}
              </div>
              <div className="oi-node-click-item">
                <NorthMenu application={application} />
              </div>
            </div>
            <Link
              to={`/north/${application.id}`}
              className="w-100 text-decoration-none text-muted flex-grow-1"
            >
              <div className="d-flex flex-column h-100 justify-content-center py-2 oi-node-click-item">
                <div>
                  {application.api}
                </div>
              </div>
            </Link>
            <div className="oi-node-footer">
              <OIbCheckBox
                name={`north.applications.${applications.findIndex(
                  (element) => element.id === application.id,
                )}.enabled`}
                defaultValue={false}
                value={application.enabled}
                onChange={onChange}
                switchButton
              />
            </div>
          </div>),
      },

      // position the node with an offset to center and then an offset for each node
      position: {
        x: 620 - (100 * Math.min(applications.length, 5)) + (indexNorth % 5) * 200,
        y: 150 * Math.trunc(indexNorth / 5),
      },
    }
  ))
  const northLinks = applications.map((application) => (
    {
      id: `${application.id}-engine`,
      source: 'engine',
      target: application.id,
      animated: true,
      type: 'default',
      arrowHeadType: 'arrow',
      isHidden: !application.enabled,
    }
  ))

  const southNodes = dataSources.map((dataSource, indexSouth) => (
    {
      id: dataSource.id,
      type: 'input',
      sourcePosition: 'top',
      style: {
        background: NODE_BACKGROUND_COLOR,
        border: NODE_BORDER,
        width: 180,
        height: 130,
        padding: 0,
        borderRadius: 5,
      },
      data: {
        label: (
          <div className="d-flex flex-column h-100 w-100">
            <div className="w-100 d-flex flex-row justify-content-between align-items-center p-1 oi-node-header">
              <img
                src={`${imageCategories[ProtocolSchemas[dataSource.protocol].category].image}` ?? imageCategories.Default.image}
                alt="logo"
                height="25px"
              />
              <div id={`title-${dataSource.name}`} className="oi-node-title">
                {dataSource.name}
              </div>
              <div
                className="oi-node-click-item"
              >
                <SouthMenu dataSource={dataSource} />
              </div>
            </div>
            <Link
              to={`/south/${dataSource.id}`}
              className="w-100 text-decoration-none text-muted flex-grow-1"
            >
              <div className="d-flex flex-column h-100 justify-content-center py-2 oi-node-click-item">
                <div>
                  {dataSource.protocol}
                </div>
                {dataSource.points?.length > 0 ? (
                  <div>
                    {`Points (${dataSource.points.length})`}
                  </div>
                ) : null}
              </div>
            </Link>
            <div className="oi-node-footer">
              <OIbCheckBox
                name={`south.dataSources.${dataSources.findIndex(
                  (element) => element.id === dataSource.id,
                )}.enabled`}
                defaultValue={false}
                value={dataSource.enabled}
                onChange={onChange}
                switchButton
              />
            </div>
          </div>
        ),
      },
      // position the node with an offset to center and then an offset for each node
      // 5 per line max => potentially render on several lines with y
      position: {
        x: 620 - (100 * Math.min(dataSources.length, 5)) + (indexSouth % 5) * 200,
        y: 190 + 150 * Math.trunc(indexSouth / 5) + 150 * (Math.trunc((applications.length - 1) / 5) + 1),
      },
    }
  ))
  const southLinks = dataSources.map((dataSource) => (
    {
      id: `${dataSource.id}-engine`,
      target: 'engine',
      source: dataSource.id,
      animated: true,
      type: 'default',
      arrowHeadType: 'arrow',
      isHidden: !dataSource.enabled,
    }
  ))

  const elements = [
    ...northNodes,
    ...northLinks,
    {
      id: 'engine',
      data: {
        label: (
          <div className="d-flex flex-column h-100 w-100">
            <div className="w-100 d-flex flex-row justify-content-between align-items-center p-1 oi-node-header">
              <div />
              <div className="oi-node-title">
                {`Engine ${engineName}`}
              </div>
              <div className="oi-node-click-item">
                <EngineMenu onRestart={onRestart} onShutdown={onShutdown} />
              </div>
            </div>
            <Link
              to="/engine"
              className="w-100 text-decoration-none text-muted flex-grow-1"
            >
              <div className="d-flex flex-column h-100 justify-content-between py-2 oi-node-click-item">
                {Object.entries(connectorData)
                  .filter(([key]) => key === 'Up time' || key === 'Global memory usage' || key === 'CPU usage')
                  .map(([key, value]) => (
                    <div key={key}>
                      <b className="mr-2">
                        {key}
                        :
                      </b>
                      <span>{value}</span>
                    </div>
                  ))}
              </div>
            </Link>
          </div>
        ),
      },
      position: {
        x: 100,
        y: 20 + 150 * (Math.trunc((applications.length - 1) / 5) + 1),
      },
      targetPosition: 'bottom',
      sourcePosition: 'top',
      style: {
        background: NODE_BACKGROUND_COLOR,
        border: NODE_BORDER,
        width: 1020,
        height: 130,
        padding: 0,
        borderRadius: 5,
      },
    },
    ...southNodes,
    ...southLinks,
  ]

  return (
    <Container>
      <div style={{
        height: 410 + 150 * (Math.trunc((northNodes.length - 1) / 5) + 1) + 150 * (Math.trunc((southNodes.length - 1) / 5) + 1),
        width: '100%',
      }}
      >
        <ReactFlow
          elements={elements}
          zoomOnScroll={false}
          zoomOnDoubleClick={false}
          selectNodesOnDrag={false}
          paneMoveable={false}
          preventScrolling={false}
          nodesConnectable={false}
          elementsSelectable={false}
          nodesDraggable={false}
          onLoad={onLoad}
        />
      </div>
    </Container>
  )
}

NodeView.propTypes = {
  onRestart: PropTypes.func.isRequired,
  onShutdown: PropTypes.func.isRequired,
}

export default NodeView
