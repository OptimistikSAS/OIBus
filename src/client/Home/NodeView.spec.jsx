/**
 * @jest-environment jsdom
 */
import React from 'react'
import ReactDOM from 'react-dom'
import { act } from 'react-dom/test-utils'
import NodeView from './NodeView.jsx'
import newConfig from '../../../tests/testConfig'

const dispatchNewConfig = jest.fn()

global.EventSource = class {
  constructor() {
    this.close = () => {}
  }
}

// ReactFlow does not seem to be working with jest.
// so we have to mock this component
jest.mock('../../../node_modules/react-flow-renderer/dist/esm/index.js', () => () => ('ReactFlow'))

let container

beforeEach(() => {
  React.useContext = jest.fn().mockReturnValue({
    newConfig,
    dispatchNewConfig,
  })
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

// sample status (object returned by Server to give various informations on the behavior)
const status = { version: 'x.x.x' }

React.useContext = jest.fn().mockReturnValue({ newConfig })
describe('NodeView', () => {
  test('display main page based on config', async () => {
    act(() => {
      ReactDOM.render(
        <NodeView status={status} onRestart={() => true} onShutdown={() => true} />,
        container,
      )
    })
    expect(container).toMatchSnapshot()

    React.useContext = jest.fn().mockReturnValue({
      newConfig: null,
      dispatchNewConfig,
    })

    act(() => {
      ReactDOM.render(
        <NodeView status={status} onRestart={() => true} onShutdown={() => true} />,
        container,
      )
    })
    expect(container).toMatchSnapshot()

    React.useContext = jest.fn().mockReturnValue({
      newConfig: {},
      dispatchNewConfig,
    })

    act(() => {
      ReactDOM.render(
        <NodeView status={status} onRestart={() => true} onShutdown={() => true} />,
        container,
      )
    })
    expect(container).toMatchSnapshot()

    React.useContext = jest.fn().mockReturnValue({
      newConfig: { south: {}, north: {} },
      dispatchNewConfig,
    })

    act(() => {
      ReactDOM.render(
        <NodeView status={status} onRestart={() => true} onShutdown={() => true} />,
        container,
      )
    })
    expect(container).toMatchSnapshot()

    React.useContext = jest.fn().mockReturnValue({
      newConfig: { south: { dataSources: [] }, north: { applications: [] } },
      dispatchNewConfig,
    })

    act(() => {
      ReactDOM.render(
        <NodeView status={status} onRestart={() => true} onShutdown={() => true} />,
        container,
      )
    })
    expect(container).toMatchSnapshot()
  })
})
