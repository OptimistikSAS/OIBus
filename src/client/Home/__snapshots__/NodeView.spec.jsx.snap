// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeView display main page based on config 1`] = `
<div>
  <div
    class="container"
  >
    <div
      style="height: 710px; width: 100%; padding-top: 2rem;"
    >
      ReactFlow
    </div>
  </div>
</div>
`;

exports[`NodeView display main page based on config 2`] = `
<div>
  <div
    class="container"
  >
    <div
      style="height: 710px; width: 100%; padding-top: 2rem;"
    >
      ReactFlow
    </div>
  </div>
</div>
`;

exports[`NodeView display main page based on config 3`] = `
<div>
  <div
    class="container"
  >
    <div
      style="height: 710px; width: 100%; padding-top: 2rem;"
    >
      ReactFlow
    </div>
  </div>
</div>
`;

exports[`NodeView display main page based on config 4`] = `
<div>
  <div
    class="container"
  >
    <div
      style="height: 710px; width: 100%; padding-top: 2rem;"
    >
      ReactFlow
    </div>
  </div>
</div>
`;
